intellij/src/saros/intellij/eventhandler/filesystem/LocalFilesystemModificationHandler.java

```diff
@@ -367,35 +368,103 @@ private void generateResourceDeletionActivity(@NotNull VirtualFileEvent virtualF
       LOG.trace("Reacting before resource deletion: " + deletedVirtualFile);
     }
 
-    SPath path = VirtualFileConverter.convertToSPath(project, deletedVirtualFile);
+    if (deletedVirtualFile.isDirectory()) {
+      generateFolderDeletion(deletedVirtualFile);
+    } else {
+      generateFileDeletion(deletedVirtualFile);
+    }
+  }
+
+  /**
+   * Generates and dispatches a folder deletion activity for the deleted folder.
+   *
+   * @param deletedFolder the folder that was deleted
+   */
+  private void generateFolderDeletion(@NotNull VirtualFile deletedFolder) {
+    SPath path = VirtualFileConverter.convertToSPath(project, deletedFolder);
 
     if (path == null || !session.isShared(path.getResource())) {
       if (LOG.isTraceEnabled()) {
-        LOG.trace("Ignoring non-shared resource deletion: " + deletedVirtualFile);
+        LOG.trace("Ignoring non-shared folder deletion: " + deletedFolder);
       }
 
       return;
     }
 
     User user = session.getLocalUser();
 
-    IActivity activity;
+    Deque<IActivity> queuedDeletionActivities = new ConcurrentLinkedDeque<>();
 
-    if (deletedVirtualFile.isDirectory()) {
-      // TODO create deletion activities for child resources
-      // TODO clean up editor pool and annotations for child resources
-      activity = new FolderDeletedActivity(user, path);
+    VirtualFileFilter virtualFileFilter = getVirtualFileFilter(deletedFolder);
 
-    } else {
-      activity =
-          new FileActivity(
-              user, Type.REMOVED, FileActivity.Purpose.ACTIVITY, path, null, null, null);
+    /*
+     * Defines the actions executed on the base directory and every valid
+     * child resource (defined through the virtualFileFilter).
+     *
+     * Returns whether the content iteration should be continued.
+     */
+    ContentIterator contentIterator =
+        fileOrDir -> {
+          if (!fileOrDir.isDirectory()) {
+            generateFileDeletion(fileOrDir);
+
+            return true;
+          }
+
+          SPath folderPath = VirtualFileConverter.convertToSPath(project, fileOrDir);
+
+          if (folderPath == null) {
+            LOG.debug("Skipping deleted folder as no SPath could be obtained " + fileOrDir);
+
+            return true;
+          }
+
+          IActivity newFolderDeletedActivity = new FolderDeletedActivity(user, folderPath);
+
+          queuedDeletionActivities.addFirst(newFolderDeletedActivity);
+
+          return true;
+        };
 
-      editorManager.removeAllEditorsForPath(path);
+    /*
+     * Calls the above defined contentIterator on the base directory and
+     * every contained resource.
+     * Directories are only stepped into if they match the above defined
+     * virtualFileFilter. This also applies to the base directory.
+     */
+    VfsUtilCore.iterateChildrenRecursively(deletedFolder, virtualFileFilter, contentIterator);
+
+    while (!queuedDeletionActivities.isEmpty()) {
+      dispatchActivity(queuedDeletionActivities.pop());
```

Comment:
Hmm I was going to ask where the activities are sorted until I realized that the activitied are fired during the iteration. Not that easy to spot.

Reply:
I didn't see a reason to also queue file deletions. They are guaranteed to be leaves, meaning they can be processed by the other participants directly without running the risk of deleting child resources in the process.

I extended the javadoc to include information on how the activity order is handled.